boot/freeldr/freeldr/arch/realmode/helpers_all.inc

```diff
@@ -0,0 +1,155 @@
+
+cs_writestr:
+    /* Save necessary registers */
+    push ax
+    push ds
+
+    /* Prepare ds before writestr call */
+    mov ax, cs
+    mov ds, ax
+
+    /* Display the message */
+    call writestr
+
+    /* Restore necessary registers and return */
+    pop ds
+    pop ax
+    ret
+
+/* This function will Update BSS_CurrentBaseAddress, instruction pointer, code segment and relocate if necessary */
+RelocateFreeLdr:
+    /* Clear necessary registers */
+    xor eax, eax
+    xor ebx, ebx
+    xor ecx, ecx
+
+    /* Call the sub function to get the base address */
+    call .SubCall
+.SubCall:
+    /* Pop the return address into bx. Since the return address is consumed,
+     * this function does not return to UpdateCurrentCodeSegment */
+    pop bx
+
+    /* Calculate the base address by adjusting the offset */
+    sub bx, offset .SubCall
+
+    /* Load the current code segment into eax */
+    mov ax, cs
+
+    /* Shift the segment address to get its physical address (real mode calculation) */
+    shl eax, 4
+
+    /* Add the physical segment base to the relative base address */
+    add eax, ebx
+
+    /* Store the calculated base address (low 32 bits) */
+    mov dword ptr ds:[BSS_CurrentBaseAddress], eax
+
+.FixBaseAddress:
+    /* Don't relocate if it's not necessary */
+    cmp eax, FREELDR_BASE
+    jz .NotNecessaryRelocation
+
+    /* Setup source segment */
+    shr eax, 4
+    mov ds, ax
+
+    /* Move the realmode FREELDR code to TEMPCODE16_BASE */
+    mov di, TEMPCODE16_BASE
+    xor si, si
+    mov cx, FREELDR_PE_BASE - FREELDR_BASE
+    rep movsb
+
+    ljmp16 TEMPCODE16_BASE / 16, .FixBaseAddress2
+.FixBaseAddress2:
+    /* Prepare the registers for relocation */
+    mov cx, (MEMORY_MARGIN - FREELDR_BASE) / 16
+    mov eax, dword ptr es:[BSS_CurrentBaseAddress]
+    shr eax, 4
+    mov ds, ax
+    mov ax, FREELDR_BASE / 16
+    mov es, ax
+
+    /* If current base address is lower than the final one then just copy it backwards */
+    cmp dword ptr fs:[BSS_CurrentBaseAddress], FREELDR_BASE
+    jb .PreFixLoopInverted
+.FixLoop:
+    /* Save CX */
+    push cx
+
+    /* Copy 16 bytes from source to destination */
+    xor di, di
+    xor si, si
+    mov cx, HEX(10)
+    rep movsb
+
+    /* Restore CX */
+    pop cx
+
+    /* Increment DS */
+    mov ax, ds
+    inc ax
+    mov ds, ax
+
+    /* Increment ES */
+    mov ax, es
+    inc ax
+    mov es, ax
+
+    /* Repeat the loop while decrementing cx */
+    loop .FixLoop
+    jmp .FinishedRelocation
+.PreFixLoopInverted:
+    /* Move DS to the end */
+    mov ax, ds
+    add ax, cx
+    mov ds, ax
+
+    /* Move ES to the end */
+    mov ax, es
+    add ax, cx
+    mov es, ax
+.FixLoopInverted:
+    /* Decrement DS */
+    mov ax, ds
+    dec ax
+    mov ds, ax
+
+    /* Decrement ES */
+    mov ax, es
+    dec ax
+    mov es, ax
+
+    /* Save CX */
+    push cx
+
+    /* Copy 16 bytes from source to destination */
+    xor di, di
+    xor si, si
+    mov cx, HEX(10)
```

Comment:
If you set the direction flag and mov the size into `si` and `di` you can do large (4K) movs here, too.